janitor/functions/groupby_topk.py

```diff
@@ -41,59 +44,78 @@ def groupby_topk(
-        input DataFrame `df`.
-    :param k: Number of top rows to return from each group after sorting.
-    :param sort_values_kwargs: Arguments to be passed to sort_values function.
-    :returns: A pandas DataFrame with top `k` rows that are grouped by
-        `groupby_column_name` column with each group sorted along the
-        column `sort_column_name`.
+    :param by: Column name(s) to group input DataFrame `df` by.
+    :param column: Name of the column that determines `k` rows
+        to return.
+    :param k: Number of top rows to return for each group.
+    :param dropna: If `True`, and `NA` values exist in `by`, the `NA`
+        values are not used in the groupby computation to get the relevant
+        `k` rows. If `False`, and `NA` values exist in `by`, then the `NA`
+        values are used in the groupby computation to get the relevant
+        `k` rows. The default is `True`.
+    :param ascending: Default is `True`. If `True`, the smallest top `k` rows,
+        determined by `column` are returned; if `False, the largest top `k` rows,
+        determined by `column` are returned.
+    :param ignore_index: Default `True`. If `True`,
+        the original index is ignored. If `False`, the original index
+        for the top `k` rows is retained.
+    :returns: A pandas DataFrame with top `k` rows per `column`, grouped by `by`.
     :raises ValueError: if `k` is less than 1.
-    :raises ValueError: if `groupby_column_name` not in DataFrame `df`.
-    :raises ValueError: if `sort_column_name` not in DataFrame `df`.
-    :raises KeyError: if `inplace:True` is present in `sort_values_kwargs`.
     """  # noqa: E501
 
-    # Convert the default sort_values_kwargs from None to empty Dict
-    sort_values_kwargs = sort_values_kwargs or {}
+    # TODO: allow multiple columns for `columns`
+    # when DataFrameGroupBy.nlargest/nsmallest is implemented
+    if isinstance(by, Hashable):
+        by = [by]
+
+    check("by", by, [Hashable, list])
 
-    # Check if groupby_column_name and sort_column_name exists in the DataFrame
-    check_column(df, [groupby_column_name, sort_column_name])
+    check_column(df, [column])
+    check_column(df, by)
 
-    # Check if k is greater than 0.
     if k < 1:
         raise ValueError(
-            "Numbers of rows per group to be returned must be greater than 0."
+            "Numbers of rows per group "
+            "to be returned must be greater than 0."
         )
 
-    # Check if inplace:True in sort values kwargs because it returns None
-    if (
-        "inplace" in sort_values_kwargs.keys()
-        and sort_values_kwargs["inplace"]
-    ):
-        raise KeyError("Cannot use `inplace=True` in `sort_values_kwargs`.")
+    indices = df.groupby(by=by, dropna=dropna, sort=False, observed=True)
+    indices = indices[column]
+
+    if ascending:
+        indices = indices.nsmallest(n=k)
+    else:
+        indices = indices.nlargest(n=k)
```

Comment:
This isn't, strictly speaking, equivalent to the initial implementation involving sorting and taking the head/tail though.

`nsmallest`/`nlargest` don't work with string / object types, for example.
So at the moment, if `column` passed in is of a non-numeric dtype, `pandas` will raise an error. (It's a limitation imposed on by pandas, but I'm not quite sure why.)

Reply:
fair point ... based on this, it is fair to abandon this PR

Reply:
Before we close the PR, it may be worth thinking about the range of use cases. 

`groupby_topk` probably implies a numeric ordering, rather than lexicographic ordering, right? 

If so, then I think the implementation here, while more constrained in scope, fulfills the intent of the function more closely, and we can close out the PR by merging.

If not, then let's close the PR without merging.

Reply:
One option could be to put a check ... If it is a numeric column use nlargest/smallest, else sort?

Reply:
re. the scope of the `groupby_topk` function, there is kind of a gray area given the way the function is named. I'll be okay with going with either interpretation (only allowing numeric ordering, OR allowing generic ordering which includes lexicographical or even custom ordering).

I'm also not in favour of closing the PR, I think there is value with what Sam is proposing here.

I'm only bringing this up in the review because, given the previous implementation allowed for generic ordering, there might be some users making use of this behaviour. If we decide to go with purely numeric ordering, we need to give ample warning to the users (in the CHANGELOG & in the docstring).

That said, I would vote for Sam's first proposal (if numeric, then use nlargest/smallest; while having `groupby_topk` still supporting generic ordering for other types, via the old implementation).

Reply:
@samukweku @thatlittleboy great points. Yeah, I agree with @thatlittleboy also, that we include an if/else block to call on the appropriate chunk of code based on the `dtype` of the column being sorted. @samukweku, do you know if it would be too much work to reinstate the original code for the non-numerical sorting use case?

Reply:
I could do that. One thing with the original code though is that the grouping column becomes the index as well; I propose that the dataframe index remains as is and the grouper stays within the dataframe

Reply:
Yes, that sounds like a reasonable change to make, I think. Let's go with that.